torch/library.h

```diff
@@ -486,6 +554,13 @@ class TORCH_API Library final {
     return _impl(name, std::move(f));
   }
 
+  // See Note [Plumbing Keys Through The Dispatcher]
+  template <typename Name, typename Func>
+  Library& impl_withKeys(Name name, Func&& raw_f) & {
+    CppFunction f(CppFunction::WithKeys::WITH_KEYS, std::forward<Func>(raw_f));
+    return _impl(name, std::move(f));
+  }
```

Comment:
I think there's a way you can get this to work without having to replicate all the functions here with `_withKeys`, at the cost of a slightly more complicated registration invocation. First, this comment in this file is relevant:

```
  // Some notes about the API design here.  We had the following constraints:
  //  
  //  - We need to support multiple "types" of arguments for schema and
  //    functions (e.g., unnamed lambda types, regular functions, const char*,
  //    fully instantiated schemas)
  //  - We don't want to write exponentially many overloads
  //  - We don't want to rely on implicit conversion to a common type,
  //    because the C++ compiler will only be willing to do a single
  //    implicit conversion (reducing the set of valid types which you
  //    can invoke with); also error messages are worse when an implicit
  //    conversion is not selected (as the compiler will not explain
  //    why it didn't select an implicit conversion; this is different
  //    from overloads where it will explain each candidate overload and
  //    why it didn't apply)
  //
  // To solve all of these constraints at the same time, we use a trick taken
  // from the pybind11 library: template over the argument in the user visible
  // API, and inside of the templated function explicitly call an overloaded
  // function to resolve the argument to a real type.  You get the good error
  // messages from overloads, but at the same time you only need to write the
  // overload for any given argument type once.
```

So the idea is that when you want to pass a with keys kernel function, you should simply pass in a `Func` that directly "encodes" the fact that this is a with keys function. There are several ways you could do this, but the most straightforward is to directly pass an already constructed CppFunction that was constructed `WITH_KEYS`. Then the normal `impl` function works.

Reply:
Thanks, I'll try that. I think I should be able to add CppFunction constructor overloads to differentiate those two cases (in which case I wouldn't need the enum at all).